google/resource_sql_source_representation_instance.go

```diff
@@ -0,0 +1,227 @@
+				ForceNew: true,
+			},
+
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"self_link": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceSqlSourceRepresentationInstanceCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	var name string
+	if v, ok := d.GetOk("name"); ok {
+		name = v.(string)
+	} else {
+		name = resource.UniqueId()
+	}
+	d.Set("name", name)
+
+	host := d.Get("host").(string)
+	port := d.Get("port").(int)
+	hostPort := fmt.Sprintf("%s:%d", host, port)
+
+	instance := &sqladmin.DatabaseInstance{
+		Name:            name,
+		Region:          region,
+		DatabaseVersion: d.Get("database_version").(string),
+		OnPremisesConfiguration: &sqladmin.OnPremisesConfiguration{
+			HostPort: hostPort,
+		},
+	}
+
+	backoff := time.Second
+	var op *sqladmin.Operation
+	for {
+		op, err = config.clientSqlAdmin.Instances.Insert(project, instance).Do()
+		if err == nil {
+			break
+		}
+
+		// When deleting and recreating a source representation instance, the (re)create operation fails with an invalidState
+		// for a very short window after the deletion succeeds. This is easily fixed by retrying when encountering that
+		// error.
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 409 && strings.Contains(gerr.Body, "invalidState") {
+			log.Printf("[DEBUG]: Got an invalid state error, retrying after %s\n", backoff)
+			time.Sleep(backoff)
+			backoff = backoff * 2
+			if backoff > 30*time.Second {
+				return errwrap.Wrapf(fmt.Sprintf("Error, constantly failing to create source representation instance %s. Too many invalid state errors. Latest error {{err}}", instance.Name), err)
+			}
+		} else {
+			return errwrap.Wrapf(fmt.Sprintf("Failed to create source representation instance %s: {{err}}", instance.Name), err)
+		}
+	}
+
+	d.SetId(instance.Name)
+
+	err = sqladminOperationWaitTime(config, op, project, "Create Source Representation Instance", int(d.Timeout(schema.TimeoutCreate).Minutes()))
+	if err != nil {
+		d.SetId("")
+		return err
+	}
+
+	return resourceSqlSourceRepresentationInstanceRead(d, meta)
+}
+
+func resourceSqlSourceRepresentationInstanceRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	instance, err := config.clientSqlAdmin.Instances.Get(project, d.Id()).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("SQL Source Representation Instance %q", d.Get("name").(string)))
+	}
+
+	d.Set("name", instance.Name)
+	d.Set("region", instance.Region)
+	d.Set("database_version", instance.DatabaseVersion)
+	hostPort := strings.Split(instance.OnPremisesConfiguration.HostPort, ":")
```

Comment:
Since instance.OnPremisesConfiguration could be nil, we should probably add a nil check here. Please also add a check on the split string to make sure it was %s:%s